### PR TITLE
Serialize new function type representation

### PR DESCRIPTION
**Pull request by slavapestov**
Stop serializing the input tuple as a single type, and start serializing individual parameters instead. This is a nice simplification because we no longer need to serialize TupleType and ParenType with parameter flags, and InOutType is now never serialized at all.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test compiler performance

**Review by CodaFi — APPROVED** [review]
You’re on fire tonight.

File: lib/Serialization/Serialization.cpp [hunk 1]
```diff
@@ -3782,29 +3776,42 @@ void Serializer::writeType(Type ty) {
...
+      assert(!fnTy->isNoEscape());
+
+      auto *genericSig = cast<GenericFunctionType>(fnTy)->getGenericSignature();
+      unsigned abbrCode = DeclTypeAbbrCodes[GenericFunctionTypeLayout::Code];
+      GenericFunctionTypeLayout::emitRecord(Out, ScratchRecord, abbrCode,
+              addTypeRef(fnTy->getResult()),
+              getRawStableFunctionTypeRepresentation(fnTy->getRepresentation()),
+              fnTy->throws(),
+              addGenericSignatureRef(genericSig));
+    }
```
Comment (jrose-apple) [inline lib/Serialization/Serialization.cpp]: This `if` is making me think we should just use the same layout for both, and put a 0 ID in for non-generic function types.

**Review by jrose-apple — COMMENTED** [review]
Looks great. I originally thought `computeType` had to calculate the generic signature anew, but that's clearly not the case.

**Comment by slavapestov** [comment]
The source compatibility failure is from attempting to serialize a tuple type with flags set. This only happens because we still “maximally reabstract” function values in SIL into a form where they take a single tuple argument, for example when you store one in an array. I was hoping to put off changing that for a while longer, but this is where all this was leading up to anyway, and now the constraint solver no longer allows such function conversions so there’s no reason for SIL to do this either.

**Comment by slavapestov** [comment]
The source compat issue should be fixed by https://github.com/apple/swift/pull/18659.

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Pull request merged**